src/material/form-field/src/formly.attributes.ts

```diff
@@ -0,0 +1,19 @@
+import { Directive, Input } from '@angular/core';
+import { FormlyAttributes, FormlyFieldConfig } from '@ngx-formly/core';
+
+@Directive({
+  selector: '[formlyMatAttributes]',
+  host: {
+    '[attr.name]': 'field.name',
+    '[attr.step]': 'to.step',
+
+    '(keyup)': 'to.keyup && to.keyup(field, $event)',
+    '(keydown)': 'to.keydown && to.keydown(field, $event)',
+    '(click)': 'to.click && to.click(field, $event)',
+    '(change)': 'to.change && to.change(field, $event)',
+    '(keypress)': 'to.keypress && to.keypress(field, $event)',
+  },
+})
+export class FormlyMatAttributes extends FormlyAttributes {
```

Comment:
I would prefer to keep `formlyAttributes` (we can adjust it to make it  compatible with material if necessary), since I'm thinking that most the material issues would be gone on `ivy` implementation,

Reply:
In fact I tried to set all attributes to `null` in material use case, but if any of mentioned attributes is present in `host:`, certain components will have some edge case issues.

Example:
- Add `'[attr.tabindex]': 'to.tabindex || 0',` back to `formlyMatAttributes`
- Run demo, navigate to Material and scroll to checkbox
- Set checkbox to `false` value
-- Now you should be getting `mat-error`, but you won't

---

Probably the only way to make this work is to add attributes with `Renderer2` and add an option to disable it for Material?

Reply:
I am just trying it out `Renderer2` and it seems to be working fine, tests are passing as well.

---

Injection:
```
@Optional() @Attribute('material') private isMaterial: boolean,
```

In `ngOnChanges` livecycle hook:
```
if (!this.isMaterial) {
  this.renderer.setAttribute(this.elementRef.nativeElement, 'placeholder', this.to.placeholder);
    this.renderer.setAttribute(this.elementRef.nativeElement, 'tabindex', `${this.to.tabindex || 0}`);
  this.renderer.setAttribute(this.elementRef.nativeElement, 'readonly', this.to.readonly);
}
```

---

I am however not happy with `material` attribute..Do you have any other idea how to detect if directive is used within a material?

One option would be to again extend the class and set the property `isMaterial` to `true`, by default it's false...don't like it either.

Reply:
can you try adding those attributes as inputs:
```ts
@Directive({
  selector: '[formlyAttributes]',
  host: {
    '[attr.tabindex]': 'tabindex ? tabindex : to.tabindex',
  },
})
export class FormlyAttributes implements OnChanges {
  @Input() tabindex;
}
```

Reply:
@aitboudad brilliant, should have think of that first.
I would however prefer the following  syntax: `'[attr.placeholder]': 'placeholder || to.placeholder',`.

What do you think?

Reply:
seems better :)

Reply:
Will do a bit of testing over the known issues of material and then will push.

Reply:
Just realised I forgot to replace `formlyMatAttributes` with `formlyAttributes`, so I was basically testing the current workaround. Sorry for that.

So the following code, sadly does not work:
```
@Directive({
  selector: '[formlyAttributes]',
  host: {
    '[attr.name]': 'field.name',
    '[attr.placeholder]': 'placeholder || to.placeholder',
    '[attr.tabindex]': 'tabindex || to.tabindex || 0',
    '[attr.readonly]': 'readonly || to.readonly',
    '[attr.step]': 'to.step',

    // ...
  },
})
export class FormlyAttributes implements OnChanges {
  @Input('formlyAttributes') field: FormlyFieldConfig;
  @Input('placeholder') placeholder: number;
  @Input('tabindex') tabindex: number;
  @Input('readonly') readonly: number;

  // ...
```

Seems like putting the attribute like that on some material components really brings in hard to notice side effects.

I will try to find a cleaner solution than `@Attribute` suggested above.

Reply:
Ok I just figured that adding the attributes using `Renderer2` does not trigger any of the edge cases I was able to indicate earlier. So I would suggest to remove attributes from `host:` and instead always add them with `Renderer2`. This I don't consider an ugly solution, since we use `Renderer2` already anyway.

Like this:
```

if (changes.field) {
  const fieldChanges = changes.field;

  this.renderer.setAttribute(this.elementRef.nativeElement, 'id', this.field.id);
  this.renderer.setAttribute(this.elementRef.nativeElement, 'placeholder', this.to.placeholder);
  this.renderer.setAttribute(this.elementRef.nativeElement, 'tabindex', `${this.to.tabindex || 0}`);
  this.renderer.setAttribute(this.elementRef.nativeElement, 'readonly', `${this.to.readonly}`);
```

> _If you wish to reproduce it, serve the demo with `aot` and try to put a checkbox  example in material UI to false - error should appear_

Are you ok with this solution?

Reply:
Yes, that's true. Basically the changes to `templateOptions` from the outside would not be  reflected on a controls, which people might find strange and file another bug report.

What about ngDoCheck  live cycle hook with a dirty checking?

Reply:
I'd prefer to wait until checking on my side tomorrow, to see if there is a simple way

Reply:
Deal :+1:
Will throw the working code with `ngDoCheck` here anyway:

```
ngDoCheck() {
  if (this.placeholder !== this.to.placeholder) {
    this.renderer.setAttribute(this.elementRef.nativeElement, 'placeholder', this.to.placeholder);
    this.placeholder = this.to.placeholder;
  }

  if (this.tabindex !== this.to.tabindex) {
    this.renderer.setAttribute(this.elementRef.nativeElement, 'tabindex', `${this.to.tabindex || 0}`);
    this.tabindex = this.to.tabindex;
  }

  if (this.readonly !== this.to.readonly) {
    this.renderer.setAttribute(this.elementRef.nativeElement, 'readonly', `${this.to.readonly}`);
    this.readonly = this.to.readonly;
  }
}
```

Reply:
@aitboudad found any better way so far?

Reply:
nop, we can go with that approach for now.

Reply:
Ok, I will push my change then 👍